### PR TITLE
Implement exponential backoff

### PR DESCRIPTION
**Pull request by shantanuraj**
Please remember the [Contributing Guidelines](https://github.com/toggl/toggl-button/blob/master/.github/CONTRIBUTING.md) :heart:

## :star2: What does this PR do?

- Max limit is an hour
- No limits on retries

## :bug: Recommendations for testing

All changes should be tested across Chrome and Firefox.

### CODE REVIEW

File: src/scripts/background.js [hunk 1]
```diff
@@ -52,11 +69,7 @@ window.TogglButton = {
     socket: null,
     maxRetryCount: 30,
     retryCount: 0,
-    retryInterval: 5,
-    get retryEnabled () {
-      return false;
-      // return TogglButton.websocket.retryCount < TogglButton.websocket.maxRetryCount;
-    }
+    retryInterval: 5
```
Comment (leafi) [inline src/scripts/background.js]: In the case of the connection repeatedly working, then closing, then working, then closing..., we'd be retrying every 5s after the connection closes. IMO this is too soon.

Suggest retryInterval should be 30 seconds

File: src/scripts/background.js [hunk 2]
```diff
@@ -18,6 +19,22 @@ const shouldTriggerNotification = (state, seconds) => {
   );
 };
 
+function randomInt (min, max) {
+  return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+function jitter () {
+  // Random amount between 0-5 seconds
+  return randomInt(0, 5);
```
Comment (leafi) [inline src/scripts/background.js]: I think, given the range of users we have on different types of connection, it's already true there's about 5 seconds of jitter (though most users would probably take 0-3 seconds to connect, not 5 seconds).

What do you think about changing this range from 0..5 to 0..30, or even 0..60? IMO 5s jitter is very little

File: src/scripts/background.js (hunk 1 above)
Comment (shantanuraj) [reply src/scripts/background.js]: It is currently 5 + rand(0, 5) in the above case

File: src/scripts/background.js [hunk 3]
```diff
@@ -18,6 +19,22 @@ const shouldTriggerNotification = (state, seconds) => {
...
+}
+
+function backoff (retryInterval, retryCount) {
+  return Math.min(
+    (retryInterval * Math.pow(2, retryCount)) + jitter(),
```
Comment (leafi) [inline src/scripts/background.js]: On 1st retry (so we've failed to connect once), retryCount is 0, correct? That's how I think the code works.. and in that case this is correct. Just want to confirm with you.

File: src/scripts/background.js (hunk 2 above)
Comment (shantanuraj) [reply src/scripts/background.js]: I'd bump either the base amount or jitter, leaning towards jitter

File: src/scripts/background.js [hunk 4]
```diff
@@ -341,21 +350,15 @@ window.TogglButton = {
...
-
-        bugsnagClient.leaveBreadcrumb(`Websocket reconnection attempt ${TogglButton.websocket.retryCount}`);
-        if (process.env.DEBUG) console.info(`Websocket reconnection attempt ${TogglButton.websocket.retryCount}`);
-
-        if (!TogglButton.websocket.retryEnabled) {
-          bugsnagClient.leaveBreadcrumb('Websocket reconnections abandoned');
-        }
-      }, retrySeconds * 1000);
-    }
+    // Retry connection, increasing the timeout each time, up to 30 minutes.
```
Comment (leafi) [inline src/scripts/background.js]: comment says 30 min but code above says ONE_HOUR

File: src/scripts/background.js (hunk 3 above)
Comment (shantanuraj) [reply src/scripts/background.js]: Yep so 5 + 0 + rand(0, 5)

File: src/scripts/background.js (hunk 4 above)
Comment (shantanuraj) [reply src/scripts/background.js]: Will fix 🙇

File: src/scripts/background.js (hunk 1 above)
Comment (leafi) [reply src/scripts/background.js]: Would still rather look at e.g. 30s + rand(0, 60s), given that base 'retryInterval: 5' is current behavior & we are demonstrably causing problems for BE

Comment (shantanuraj) [reply src/scripts/background.js]: do you think

base retry: 15 seconds
max_jitter: 60 seconds

want to keep the regular user which might get disconnected on closing laptop lid happy

so yeah worst case time in first attempt

15 + 0 = 15 seconds after they open/resume their machines

in next attempt worst case is already 30 seconds now

15 * 2 + 0 = 30 seconds

in both cases assuming jitter is randomly 0

**Comment by leafi** [comment]
Discussed numbers with shantanu in PR... awaiting new push, then will approve

**Comment by leafi** [comment]
Code LGTM - trying to install this extension now

**Pull request merged**